### ISSUE TITLE
Error: Provider produced inconsistent final plan in 0.12

### ISSUE THREAD

**Issue by rberlind** [hashicorp/terraform#19223]
When using rich type values passed into module, I got following error when doing apply:
Error: Provider produced inconsistent final plan
Additional details from error message are below.

### Terraform Version
```
$terraform -v
Terraform v0.12.0-alpha1
+ provider.aws v1.40.0-5-g4f24bc8b7
```

### Terraform Configuration Files

main.tf:
```hcl
provider "aws" {
  region = "us-west-2"
}

variable "vpc_name" {
  description = "name of the VPC"
  default = "rvt-example-vpc"
}

variable "vpc_cidr" {
  description = "CIDR for VPC"
  default = "172.16.0.0/16"
}

variable "subnet_name" {
  description = "name for subnet"
  default = "rvt-example-subnet"
}

variable "subnet_cidr" {
  description = "CIDR for subnet"
  default = "172.16.10.0/24"
}

variable "interface_ips" {
  type = list
  description = "IP for network interface"
  default = ["172.16.10.100"]
}

module "network" {
  source = "./network"
  network_config = {
    vpc_name = var.vpc_name
    vpc_cidr = var.vpc_cidr
    subnet_name = var.subnet_name
    subnet_cidr = var.subnet_cidr
  }
}

resource "aws_network_interface" "rvt" {
  subnet_id = module.network.subnet.id
  private_ips = var.interface_ips
  tags = {
    Name = "rvt-example-interface"
  }
}

resource "aws_instance" "rvt" {
  ami = "ami-22b9a343" # us-west-2
  instance_type = "t2.micro"

  tags = {
    Name = "rvt-example-instance"
  }

  network_interface {
    network_interface_id = aws_network_interface.rvt.id
    device_index = 0
  }
}

output "vpc" {
  value = module.network.vpc
}

output "subnet" {
  value = module.network.subnet
}

output "instance_private_dns" {
  value = aws_instance.rvt.private_dns
}
```
network/variables.tf
```hcl
variable "network_config" {
  type = object({
    vpc_name = string
    vpc_cidr = string
    subnet_name = string
    subnet_cidr = string
  })
}
```

network/main.tf:
```hcl
resource "aws_vpc" "my_vpc" {
  cidr_block = var.network_config.vpc_cidr
  tags = {
    Name = var.network_config.vpc_name
  }
}

resource "aws_subnet" "my_subnet" {
  vpc_id = aws_vpc.my_vpc.id
  cidr_block = var.network_config.subnet_cidr
  availability_zone = "us-west-2a"
  tags = {
    Name = var.network_config.subnet_name
  }
}
```

network/outputs.tf
```hcl
output "vpc" {
  value = aws_vpc.my_vpc
}

output "subnet" {
  value = aws_subnet.my_subnet
}
```

### Debug Output
https://gist.github.com/rberlind/468fcfcafd560c6f04e339e493fb1300

### Expected Behavior
Hoped apply would work

### Actual Behavior
When using rich type values passed into module, I got following error when doing apply:

```
Error: Provider produced inconsistent final plan

When expanding the plan for aws_network_interface.rvt to include new values
learned so far during apply, provider "aws" produced an invalid new value for
.subnet_id: was cty.NullVal(cty.String), but now
cty.StringVal("subnet-071084f85a92fb73b").

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
```

### Steps to Reproduce
1. `terraform init`
2. `terraform apply`

**Comment by rberlind** [comment]
This seems to work correctly in 0.12 alpha-2

**Comment by apparentlymart** [comment]
Thanks for confirming, @rberlind!

**Comment by jbardin** [comment]
Re-opening for now. A second `apply` will work once the module output is known, but the primary issue of an unknown module output being recorded as null is still there.

**Comment by apparentlymart** [comment]
I believe this is now fixed by #19237, but we'll hold this open for the moment until we have a fresh alpha build with providers to test against.

**Comment by jbardin** [comment]
OK, tested with master, and this is now officially closed ;)
